Maple2.Server.Game/Manager/BlackMarketManager.cs

```diff
@@ -88,16 +88,271 @@ public sealed class BlackMarketManager {
         var error = (BlackMarketError) response.Error;
         if (error != BlackMarketError.none) {
             session.Send(BlackMarketPacket.Error(error));
-            // TODO: Give item and deposit by mail.
+            CreateRefundErrorMail(item, depositFee);
             return;
         }
 
-        listings.Add(listing.Id, listing);
         session.Send(BlackMarketPacket.Add(listing));
     }
 
+    public void Remove(long listingId) {
+        using GameStorage.Request db = session.GameStorage.Context();
+
+        BlackMarketListing? listing = db.GetBlackMarketListing(listingId);
+        if (listing == null || listing.CharacterId != session.CharacterId) {
+            session.Send(BlackMarketPacket.Error(BlackMarketError.s_blackmarket_error_close));
+            return;
+        }
+
+        BlackMarketResponse response = session.World.BlackMarket(new BlackMarketRequest {
+            Remove = new BlackMarketRequest.Types.Remove {
+                ListingId = listingId,
+            },
+        });
+
+        var error = (BlackMarketError) response.Error;
+        if (error != BlackMarketError.none) {
+            session.Send(BlackMarketPacket.Error(error));
+            return;
+        }
+
+        long deposit = listing.ExpiryTime < DateTime.Now.ToEpochSeconds() ? listing.Deposit : 0;
+
+        var mail = new Mail {
+            ReceiverId = session.CharacterId,
+            Type = MailType.BlackMarketListingCancel,
+            TitleArgs = new[] {
+                ("item", $"{listing.Item.Id}"),
+            },
+            ContentArgs = new[] {
+                ("key", listing.ExpiryTime < DateTime.Now.ToEpochSeconds() ?
+                    $"{StringCode.s_blackmarket_mail_to_cancel_expired}" :
+                    $"{StringCode.s_blackmarket_mail_to_cancel_direct}"),
+                ("item", $"{listing.Item.Id}"),
+                ("str", $"{listing.Quantity}"),
+                ("money", $"{listing.Price * listing.Quantity}"),
+                ("money", $"{listing.Price}"),
+                ("money", $"{deposit}"),
+            },
+            Meso = deposit,
+        };
+
+        mail.SetTitle(StringCode.s_blackmarket_mail_to_cancel_title);
+        mail.SetContent(StringCode.s_blackmarket_mail_to_cancel_content);
+        mail.SetSenderName(StringCode.s_blackmarket_mail_to_sender);
```

Comment:
can't put these setters in the class initializer?

Reply:
it wont look as nice. This is what SetTitle, SetContent, SetSenderName all (similarly) do:

```csharp
    public void SetTitle(StringCode title) {
        Title = $"""<ms2><v key="{title.ToString()}" /></ms2>""";
    }```

Reply:
ah ok!